### PR TITLE
Extended fix OOM Issue #21634 on KerasHub side

### PR DESCRIPTION
**Pull request by praveenhosdrug123**
## Summary
Applies distributed initialization fix to model presets to resolve OOM errors during initialization of 7B+ parameter models on 8GB TPU devices. This PR updates Keras Hub model backbone and adds a dist_initializer.

## Issue
Token embedding initialization creates large arrays at time of creation, placing all weights on a single device.
Combined with forward passes during backbone initialization, this causes a 2X to 3X memory spike and triggers OOM on TPUs with limited HBM.

## Solution
Updates model backbone (Gemma, Llama, Qwen) to use dist_initializer for token embeddings, sharding weights across TPU devices during instantiation. Validated on 8-device TPU: models that previously OOM'd during backbone initialization now load successfully. 

## Reference
For memory profiling analysis, cache locality theory, validation logs and alternative solutions considered, refer to: [Doc](https://docs.google.com/document/d/1USQj3Tn1uL7iUijbPGLAsE4NiB_BDPlRRGihAgjWj_k)

Related PR: keras-team/keras#21755

## Checklist
- [o] I have added all the necessary unit tests for my change. -- I have added smoke tests since the exact test is difficult to replicate on CPU/GPU runtime. But i have validated the fix using TPUs and the validation logs are inside the document link.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

**Comment by amitsrivastava78** [comment]
@praveenhosdrug123 thanks for the PR, i understand you are sharding the tensor at the build time itself so that instead of creating a large tensor you now created sharded tensor on respective device directly.
but i am just curious about the existing flow
build() → add_weight() → assign() → _direct_assign() → jax.device_put(sharded)
so the host tensor is large one but eventually on the device/s it is creating sharded tensor and then loading to respective device, you can take an example of my colab below and explain is your PR fixing host or device memory issue or both
https://colab.research.google.com/gist/amitsrivastava78/f7f1990b154001e098c9d8cfe248e4ee/issuelargegemaamodel.ipynb

**Comment by praveenhosdrug123** [comment]
@amitsrivastava78 -- Thanks for the detailed question and the collab example!

The PR fixes the host OOM issue during the build phase.

## Current Failure Point

The model OOMs during the build phase of the Embedding layer, 
specifically during the forward pass of text embedding - before any device 
sharding logic is reached.

**Expected flow:**
build() → add_weight() → assign() → _direct_assign() → jax.device_put(sharded)

**Actual flow with large models:**
Failure happens at build() before the host array reaches the sharding logic 
in _direct_assign().

## Why This is Difficult to Reproduce on Single-RAM Systems

Testing on a single large RAM (even when logically divided into shards) 
doesn't reproduce the real memory pressure because:

- The entire array still materializes in one physical memory space
- Memory is then distributed logically, but the materialization already happened
- On real multi-device TPUs (8 devices × 8GB), each device has separate physical memory
- The build and forward passes happen at host level, causing real OOM before any 
  device sharding occurs

## Root Cause & Approach

The root cause behind the memory spike is somewhat speculative. I've documented 
the theory and entire approach in detail [here](https://docs.google.com/document/d/1USQj3Tn1uL7iUijbPGLAsE4NiB_BDPlRRGihAgjWj_k). 
Please have a look - if something doesn't sit right with you, happy to discuss 
or investigate further.

## Evidence: TPU Test Results

Error logs from TPU testing show the failure occurs specifically during build 
of the ReversibleEmbedding layer:

File "/.../keras_hub/src/layers/modeling/reversible_embedding.py", line 103, in build
super().build(inputs_shape)

**Before changes:**
[errors_during_forward_pass.txt](https://github.com/user-attachments/files/22983246/errors_during_forward_pass.txt)

**After changes:**
[using_distributed_initializers.txt](https://github.com/user-attachments/files/22983248/using_distributed_initializers.txt)

## Memory Analysis

A quick calculation reaffirms the memory profiling behavior observed:

**From successful run (with distributed initialization):**
- Individual device memory: 0.37GB per device after text embedding scaling
- Total memory across 8 devices: 2.96GB
- Applying observed peak memory factor (3x): 2.96GB × 3 = 8.88GB
- Available memory per device: 7.48GB
- **Overhead: 8.88GB - 7.48GB = 1.4GB**

**From error log (without distributed initialization):**
- Attempted allocation: 2.93GB
- Available memory: 1.63GB  
- **Overhead: 2.93GB - 1.63GB = 1.3GB**

**Conclusion:** The ~1.3-1.4GB overhead matches across both scenarios, 
confirming the ~3x peak memory factor during initialization.

When using sharding at token embedding intializer array, the memory pressure alleviates and the issue resolves.

Happy to run additional experiments or profiling if it would help!

File: keras_hub/src/utils/dist_initializer.py [hunk 1]
```diff
@@ -0,0 +1,647 @@
...
+from keras.src.initializers.random_initializers import RandomNormal
+from keras.src.initializers.random_initializers import RandomUniform
+from keras.src.initializers.random_initializers import TruncatedNormal
+from keras.src.initializers.random_initializers import VarianceScaling
+
+# Constant for token embedding path in layout map for all text based models.
+# This may need to be updated if models with different layout maps are used.
+# The failsafe is to just use non-distributed initializers if no layout is found
+
+TOKEN_EMBEDDING_PATH = "token_embedding/embeddings"
```
[the remark this answers is not in the document]
Comment (praveenhosdrug123) [reply keras_hub/src/utils/dist_initializer.py:21]: The hardcoded TOKEN_EMBEDDING_PATH is fragile. If a new model uses a different name for the token embedding layer or its weights, the distributed initialization will silently fall back to the default, non-distributed one. This could lead to unexpected OOM errors for large models, which this PR aims to fix. 
-- A warning has been added to the subsequent function being called through the distributed initialization. 

To make this more robust and easier to debug, consider adding a warning when a distribution strategy is active but the embedding path is not found in the layout map. This would alert developers if a new model is not correctly configured for distributed initialization. -- This is a good edge case for warning, i will include this in the next push. TODO : praveenhosrug123

File: keras_hub/src/utils/dist_initializer.py [hunk 2]
```diff
@@ -0,0 +1,647 @@
...
+    if len(mesh.shape) != 2 or len(tensor_layout.axes) != 2:
+        return None
+
+    mesh_dict = dict(zip(mesh.axis_names, mesh.shape))
+
+    num_shards_per_axis = tuple(
+        mesh_dict.get(axis, 1) for axis in tensor_layout.axes
+    )
+
+    return num_shards_per_axis
```
[the remark this answers is not in the document]
Comment (praveenhosdrug123) [reply keras_hub/src/utils/dist_initializer.py:83]: This is not planned for immediate future use, but i do think it is an important and counter-intuitive concept which should be included for future use. Scaling from a stats POV on distributed frameworks is still rare as a concept and would help future use cases.

**Comment by praveenhosdrug123** [comment]
Hi @amitsrivastava78,

This PR depends on keras-team/keras#21755 (Keras repo) being merged first.

Suggestion: Would it make sense for you to review #21755 on the Keras side, and have another team member review this KerasHub PR? That way both can progress in parallel.

Let me know if you need any clarification on the approach or additional profiling!

**Comment by hertschuh** [comment]
@praveenhosdrug123

Thank you for the investigation. This is indeed an issue.

Somebody on the team is working on a fix that's generally applicable to all variables so that you don't have to explicitly use the fix that you provided here.

**Comment by praveenhosdrug123** [comment]
@hertschuh - Thanks for the feedback and for taking the time to review the document.

I want to clarify the technical issue:
The OOM problem is about large contiguous memory allocation, not total parameter count. Token embeddings are the largest single array and exceed device memory during initialization, even when the full model would fit after sharding.

Thank you for the context on the general solution. A few follow-up questions to help me understand the timeline:

1. What's the expected completion date for the general fix?
2. Will it handle the edge cases mentioned in the document (interleaving, quantization, LoRA)?
3. Will it detect which variables actually need distribution?

The reason I ask: users are blocked on this today for 7B+ models on 8GB TPU devices. 
If the general fix is months out, would it make sense to:
- Merge this targeted fix as a stopgap
- Mark it deprecated once the general solution ships
- Remove it in a future release

Let me know if that's feasible.

**Comment by amitsrivastava78** [comment]
@praveenhosdrug123 thanks for your comments, sorry was held up with other things, i have checked your PR, in principle i do agree with your approach of solving the host OOM by creating sharded tensor on respective device,  however can you please explore the fix to be there only on keras rather than keras hub ? A good starting point(not the full solution) will be to make changes to keras/src/initializers/random_initializers.py and ensure that RandomNormal and RandomUniform are changed so that it can create a sharded tensor on the respective device. This would ensure that solution is truely generic
Let me know if this approach works for you as this will address the concerns raised by @hertschuh  as well.

**Comment by praveenhosdrug123** [comment]
Hi @amitsrivastava78, cc @hertschuh @divyashreepathihalli ,

I understand the preference for a core Keras solution. Let me explain why I chose the routing-based approach in Keras-Hub utils.

Technical reasoning:
1. Domain-specific problem: The OOM issue occurs specifically during Keras-Hub model loading with large embeddings. Keeping the fix scoped to Keras-Hub utils contains it to where the problem actually manifests.
2. Not all sharding is desirable: Adding layout constraints to core RandomNormal/RandomUniform would apply sharding logic universally - but this isn't always correct. Example: Gemma3 has a mixture of text-only and vision components where you explicitly DON'T want to shard embeddings at onset in case its not a text only model. A layout-constraint-only approach would shard regardless, causing incorrect behavior.
3. Controlled opt-in: The distributed initializer approach allows fine-grained control over which variables get sharded during initialization, based on model architecture requirements - not just generic layout rules.

The core issue: We need logic that says "shard THIS variable in THIS model because it's large" - not "shard all variable everywhere."

Next steps:
My understanding is @amitsrivastava78, @hertschuh, and @divyashreepathihalli  should align on what the recommended approach should be. I can see why each perspective has merit, but these are the technical constraints I encountered while working through the problem.

Happy to discuss further, but I'd suggest the team reach consensus on architectural direction first.

Thanks, 
Praveen

**Comment by hertschuh** [comment]
> 2. Not all sharding is desirable: Adding layout constraints to core RandomNormal/RandomUniform would apply sharding logic universally - but this isn't always correct. Example: Gemma3 has a mixture of text-only and vision components where you explicitly DON'T want to shard embeddings at onset in case its not a text only model. A layout-constraint-only approach would shard regardless, causing incorrect behavior.
> 3. Controlled opt-in: The distributed initializer approach allows fine-grained control over which variables get sharded during initialization, based on model architecture requirements - not just generic layout rules.

`keras.Variable` knows if there is a sharding, and knows each sharding specifically. So we're not talking about applying random sharding to all variables blindly.

Simply, this is about changing the current sequence:
```
1. call initializer (returns an array on device 0)
2. shard the returned array according to `self._layout`
```
to
```python
if self._layout:
  # initialize and shard at the same time
  value = jax.jit(outsharding=self._layout)(functool.partial(initializer, self.shape, self.dtype)()
else:
  value = initializer(self.shape, self.dtype)
```

Of course this pseudo code is neither complete nor accurate. It require modifying `JaxVariable`, adding overrides, refactoring a bit.

**Comment by praveenhosdrug123** [comment]
Understood. The consensus is to make changes in source Keras only, not KerasHub.

I've made the following changes locally to support sharding token embeddings at initialization:

Core changes in JaxVariable:

Added check_distributed_init()  and overriding the  _initialize_with_initializer method from the Common JaX variable.

check_distributed_init()  validates:

Initializer supports layout parameter
Variable is a token embedding (name == "embeddings" and path == "token_embedding/embeddings")
PartitionSpec contains actual sharding directives
Modified _initialize_with_initializer() to conditionally pass layout to the initializer when distributed init is needed

```python
def _initialize_with_initializer(self, initializer):
    self.set_tensor_layout()
    if self.check_distributed_init(initializer):
        value = self._convert_to_tensor(
            initializer(self._shape, dtype=self._dtype, layout=self._layout)
        )
    else:
        value = self._convert_to_tensor(
            initializer(self._shape, dtype=self._dtype)
        )
    self._initialize(value)
```

These changes specifically target token embedding initializers while keeping everything else constant.

Initialization flow:
JaxVariable (detects & triggers) → Backend-agnostic Random initializers (receives layout) → JAX random (partial function creation) → JAX distribution library (applies sharding)
Refactored relevant components to support this flow. 

Please let me know if this approach doesn't align well with the context of this pull request.

Validation: Token embedding initialization works well for architectures with only token embeddings as its embedding component.

Since this approach is architecture-agnostic, I'm running experiments across different model architectures to validate compositional behavior.  I'll update accordingly if any issues arise.

**Comment by praveenhosdrug123** [comment]
Good news All - Verified on different architectures through model presets.I've expanded beyond token embeddings to implement initialization for all variables. Refactored the seed management for JAX backend while maintaining reproducibility.

Validation Results
Tested across multiple model presets with full reproducibility:
Gemma 2B on TPU v2-8:

Seed 42, Run 1: loss=10.96871, acc=0.03821
Seed 42, Run 2: loss=10.96871, acc=0.03821 

Seed 456, Run 1: loss=10.9504, acc=0.04822
Seed 456, Run 2: loss=10.9504, acc=0.04822

Known limitation: I have documented the known limitations of the approach. For future improvements.
Let me know if you'd like any adjustments to the approach.

https://github.com/keras-team/keras/pull/21755 has been revised

**Comment by praveenhosdrug123** [comment]
@divyashreepathihalli -
The two CI failures (convert_gpt2_test and convert_t5gemma_test) appear to be unrelated to the current pull request. I've run these tests locally without issues.

Both architectures share a common embedding dropout layer, which feels like a red herring. My initial assessment focused on loading models from fixed presets mostly from the CausalLM genre of models and didn't dive deep into these specific architectures.

I'll investigate these architectures more thoroughly this weekend to confirm they don't impact the broader fix in [PR #21755](https://github.com/keras-team/keras/pull/21755).

**Comment by praveenhosdrug123** [comment]
Update:

Both GPT-2 and T5Gemma backbones work with distribution enabled.

Note on Vocabulary Size: The vocabulary size (50,257 for GPT-2) is not a multiple of 8, which can cause issues with token embedding on distributed devices. See: https://www.kaggle.com/models/keras/gpt2

**Comment by praveenhosdrug123** [comment]
Closing this PR, changes are made in Keras PR #21755 , Note Keras Hub model architectures are distribution agnostic.
